usersession/userd/privileged_desktop_launcher.go

```diff
@@ -0,0 +1,249 @@
+	}
+
+	// Iterate through the potential subdirectories formed by the first i elements of the desktop file ID.
+	// Maybe this is overkill: At the time of writing, the only use is in desktopFileIDToFilename() and there
+	// we're only checking dirs.SnapDesktopFilesDir (not all entries in $XDG_DATA_DIRS) and we know that snapd
+	// does not create subdirectories in that location.
+	for i := 1; i != len(splitFileId); i++ {
+		desktopFile, err := findDesktopFile(filepath.Join(baseDir, strings.Join(splitFileId[:i], "-")), splitFileId[i:])
+		if err == nil {
+			return desktopFile, err
+		}
+	}
+
+	return "", fmt.Errorf("could not find desktop file")
+}
+
+// desktopFileIDToFilename determines the path associated with a desktop file ID.
+func desktopFileIDToFilename(desktopFileID string) (string, error) {
+	// OpenDesktopEntry() currently only supports launching snap applications from
+	// desktop files in /var/lib/snapd/desktop/applications and these desktop files are
+	// written by snapd and considered safe for userd to process.
+	// Since we only access /var/lib/snapd/desktop/applications, ignore
+	// https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
+	baseDir := dirs.SnapDesktopFilesDir
+
+	desktopFile, err := findDesktopFile(baseDir, strings.Split(desktopFileID, "-"))
+
+	if err == nil {
+		return desktopFile, nil
+	}
+
+	return "", fmt.Errorf("cannot find desktop file for %q", desktopFileID)
+}
+
+// verifyDesktopFileLocation checks the desktop file location:
+// we only consider desktop files in dirs.SnapDesktopFilesDir
+func verifyDesktopFileLocation(desktopFile string) error {
+	if filepath.Clean(desktopFile) != desktopFile {
+		return fmt.Errorf("desktop file has unclean path: %q", desktopFile)
+	}
+
+	if !strings.HasPrefix(desktopFile, dirs.SnapDesktopFilesDir+"/") {
+		// We currently only support launching snap applications from desktop files in
+		// /var/lib/snapd/desktop/applications and these desktop files are written by snapd and
+		// considered safe for userd to process. If other directories are added in the future,
+		// verifyDesktopFileLocation() and parseExecCommand() may need to be updated.
+		return fmt.Errorf("internal error: only launching snap applications from /var/lib/snapd/desktop/applications is supported")
+	}
+
+	return nil
+}
+
+// readExecCommandFromDesktopFile parses the desktop file to get the Exec entry and
+// checks that the BAMF_DESKTOP_FILE_HINT is present and refers to the desktop file.
+func readExecCommandFromDesktopFile(desktopFile string) (exec string, icon string, err error) {
+	file, err := os.Open(desktopFile)
+	if err != nil {
+		return exec, icon, err
+	}
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+
+	inDesktopSection := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		if line == "[Desktop Entry]" {
+			inDesktopSection = true
+		} else if strings.HasPrefix(line, "[Desktop Action ") {
+			// maybe later we'll add support here
```

Comment:
can you change this to a TODO:

```suggestion
			// TODO: add support for desktop action sections
```

Reply:
If we don't have an implementation here, then the entire branch of the if statement could be removed/commented out in favour of the next one.